nixos/modules/services/web-apps/booklore.nix

```diff
@@ -0,0 +1,330 @@
+          Refer to the [documentation](https://github.com/adityachandelgit/BookLore?tab=readme-ov-file#-remote-authentication-trusted-header-sso-forward-auth) for options.
+          To set a database password use DATABASE_PASSWORD:
+          ```
+          { DATABASE_PASSWORD = "/path/to/secret/containing/password"; }
+          ```
+        '';
+      };
+
+      host = mkOption {
+        description = "The host BookLore binds to.";
+        default = "localhost";
+        example = "0.0.0.0";
+        type = types.str;
+      };
+
+      port = mkOption {
+        description = "The TCP port BookLore will listen on.";
+        default = 8080;
+        type = types.port;
+      };
+
+      user = mkOption {
+        description = "User account under which BookLore runs.";
+        default = "booklore";
+        type = types.str;
+      };
+
+      group = mkOption {
+        description = "Group under which BookLore runs.";
+        default = "booklore";
+        type = types.str;
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.booklore = {
+      description = "BookLore is a web app for hosting, managing, and exploring books, with support for PDFs, eBooks, reading progress, metadata, and stats.";
+
+      after = [ "network.target" ] ++ lib.optionals cfg.database.createLocally [ "mysql.service" ];
+      wantedBy = [ "multi-user.target" ] ++ lib.optionals cfg.database.createLocally [ "mysql.service" ];
+
+      environment = cfg.environment;
+
+      serviceConfig = {
+        Type = "simple";
+        User = cfg.user;
+        Group = cfg.group;
+        StateDirectory = "booklore";
+        ExecStartPre = "+${startupScript ''
+          echo "ALTER USER '${cfg.database.user}'@'localhost' IDENTIFIED BY '$DATABASE_PASSWORD';" | ${pkgs.mariadb}/bin/mariadb
+        ''}";
+        ExecStart = startupScript (lib.getExe cfg.package);
+
+        Restart = "on-failure";
+        EnvironmentFile = cfg.environmentFile;
+
+        # Hardening
+        CapabilityBoundingSet = "";
+        NoNewPrivileges = true;
+        PrivateUsers = true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        PrivateMounts = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        RestrictAddressFamilies = [
+          "AF_INET"
+          "AF_INET6"
+          "AF_UNIX"
+        ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+      };
+    };
+
+    services.booklore.environment = {
+      BOOKLORE_DATA_DIR = cfg.dataDir;
+      BOOKLORE_BOOKDROP_DIR = cfg.bookdropDir;
+      DATABASE_HOST = if cfg.database.host != null then cfg.database.host else "127.0.0.1";
+      DATABASE_PORT = if cfg.database.port != null then toString cfg.database.port else "3306";
```

Comment:
It might be better to just set these defaults in the option default instead (and disallow `null`) if they are going to be defaults anyway. It makes it easier for the downstream use to know what's going on.